pyexchange/uniswapv2.py

```diff
@@ -0,0 +1,342 @@
+
+        self.pair_address = self.get_pair_address(self.token_a.address, self.token_b.address)
+        self.is_new_pool = self.pair_address == Address("0x0000000000000000000000000000000000000000")
+        if not self.is_new_pool:
+            self.set_and_approve_pair_token(self.pair_address)
+
+        self.account_address = Address(self.web3.eth.defaultAccount)
+
+        #     TODO: Add permit support
+        # self.ec_signature_r = ec_signature_r
+            # self.ec_signature_s = ec_signature_s
+            # self.ec_signature_v = ec_signature_v
+
+    def set_and_approve_pair_token(self, pair_address: Address):
+        self._pair_contract = self._get_contract(self.web3, self.pair_abi['abi'], pair_address)
+        self.pair_token = Token('Liquidity', pair_address, 18)
+        self.approve(self.pair_token)
+
+    def get_account_token_balance(self, token: Token) -> Wad:
+        assert (isinstance(token, Token))
+
+        return token.normalize_amount(ERC20Token(web3=self.web3, address=token.address).balance_of(self.account_address))
+
+    def get_account_eth_balance(self) -> Wad:
+        return Wad.from_number(Web3.fromWei(self.web3.eth.getBalance(self.account_address.address), 'ether'))
+
+    def get_exchange_balance(self, token: Token, pair_address: Address) -> Wad:
+        assert (isinstance(token, Token))
+        assert (isinstance(pair_address, Address))
+
+        return token.normalize_amount(ERC20Token(web3=self.web3, address=token.address).balance_of(pair_address))
+
+    # retrieve exchange rate for the instance's pair token
+    def get_exchange_rate(self) -> Wad:
+        pair_address = self.get_pair_address(self.token_a.address, self.token_b.address)
+
+        token_a_reserve = self.get_exchange_balance(self.token_a, pair_address)
+        token_b_reserve = self.get_exchange_balance(self.token_b, pair_address)
+
+        return token_a_reserve / token_b_reserve
+
+    # Return the total number of liquidity tokens minted for a given pair
+    def get_total_liquidity(self) -> Wad:
+        return Wad(self._pair_contract.functions.totalSupply().call())
+
+    # Return our liquidity token balance
+    def get_current_liquidity(self) -> Wad:
+        return Wad(self._pair_contract.functions.balanceOf(self.account_address.address).call())
+
+    # Return a pools minimum liquidity token balance
+    def get_minimum_liquidity(self) -> Wad:
+        return Wad(self._pair_contract.functions.MINIMUM_LIQUIDITY(self.account_address.address).call())
+
+    def get_pair_address(self, token_a_address: Address, token_b_address: Address) -> Address:
+        assert (isinstance(token_a_address, Address))
+        assert (isinstance(token_b_address, Address))
+
+        return Address(self._factory_contract.functions.getPair(token_a_address.address, token_b_address.address).call())
+
+    # TODO: determine appropriate amount default
+    def approve(self, token: Token, amount: int = 10):
```

Comment:
Do we need a default?  I don't see an interface contract which mandates omission of the parameter.

Reply:
Not really, I was thinking about passing it through as a `kwarg`, but that doesn't seem to be necessary / will potentially add too much complexity. I've removed it.